app/actions/utils.py

```diff
@@ -1,11 +1,45 @@
 import logging
 import dateparser
 
+from app.services.state import IntegrationStateManager
+
 logger = logging.getLogger(__name__)
+state_manager = IntegrationStateManager()
+
 
+async def convert_to_er_observation(integration_id, galooli_record, reports_timezone):
+    async def save_device_state_and_build_observation():
+        # Save current device status
+        state = {"status": status}
 
-def convert_to_er_observation(galooli_record, reports_timezone):
+        await state_manager.set_state(
+            integration_id=integration_id,
+            action_id="pull_observations",
+            source_id=sensor_id,
+            state=state
+        )
 
+        localized_gps_time = reports_timezone.localize(dateparser.parse(time))
+        obs = {
+            'source': sensor_id,
+            'source_name': subject_name,
+            'subject_type': "security_vehicle",
+            'type': 'tracking-device',
+            'recorded_at': localized_gps_time.isoformat(),
+            'location': {
+                'lat': latitude,
+                'lon': longitude
+            },
+            'additional': {
+                'sensor_id': sensor_id,
+                'org_name': org_name,
+                'status': status,
+                'distance': distance,
+                'speed': speed,
+                'subject_groups': ['Vehicles', ]
+            }
+        }
+        return obs
```

Comment:
I recommend re-coding this to not create a new function on each iteration. It will be best if this function can simply convert the data to a Gundi observation regardless of what the status is.
Then let a calling function filter by status and do what is necessary to compare to the previously recorded state.